[worker] Worker performance increase (#11889)

gileri opened the pull request:
<!--
Thank you very much for your pull request to the OpenCTI project! We as a community
driven project depend on support and contributions like this!

Thus already a BIG THANK YOU upfront to you for choosing to help with your PR.
-->

### Proposed changes

* Align AMQP prefetch with requested worker thread pool size
* Removed busy polling for thread termination. Instead, the above [AMQP prefetch limits](https://www.rabbitmq.com/docs/consumer-prefetch) will limit the throughput.

Those two changes does allow having better utilization of resources, limiting the need of running multiple worker processes.

Remaining ideas to implement in another PR:

* [Raise pycti Requests module maximum connections, to allow more than 10 connections in parallel](https://docs.python-requests.org/en/latest/api/#requests.adapters.HTTPAdapter)
* Finer prefetch adjustment for additional performance

### Related issues
<!-- Please attach your PR to related issues in the Development widget on the right -->
* #4936
* https://github.com/OpenCTI-Platform/opencti/issues/11889

### Checklist

<!--
Please submit the source code in a way, where you could honestly say `This code is finished`.
If you feel that there are possibilities for improving the code quality, please do so.
By doing this, you are actively helping us to improve the quality of the entire OpenCTI project.
-->

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case (coverage and e2e)
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

<!-- _NOTE: Test coverage are, by default, mandatory. It will help us to improve stability of the platform. If you consider test are not relevant for this PR, reach out and explain why_ -->
<!-- For completed items, change [ ] to [x]. -->

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

This PR may increase the load on the OpenCTI API and backends, up to 5 times on single RabbitMQ active queue workloads (Only one queue has messages), down to ~5-10% on systems with more than 5 active queues

Fixes #11889

gileri commented:
I have tested with multiple active queue, and a worker size of 20, it works fine.

Here's the rate with, then without the PR changes (on 10 workers, pool size of 30):

<img width="432" height="204" alt="Screenshot From 2025-07-31 17-34-00" src="https://github.com/user-attachments/assets/7d3c5fd2-b889-4d2d-861e-6dfddd3504f4" />
<img width="432" height="204" alt="Screenshot From 2025-07-31 17-32-32" src="https://github.com/user-attachments/assets/294ef09c-7ea2-4043-a6c3-15a9c2bc876f" />

<img width="697" height="193" alt="Screenshot From 2025-07-31 17-32-29" src="https://github.com/user-attachments/assets/3d90607f-cc73-4f1d-be1e-6a52ed50d788" />

xfournet commented on opencti-worker/src/worker.py:
```diff
@@ -643,7 +645,8 @@ def __post_init__(self) -> None:
-        self.execution_pool = ThreadPoolExecutor(max_workers=self.opencti_pool_size)
+        global execution_pool
+        execution_pool = ThreadPoolExecutor(max_workers=self.opencti_pool_size)
         self.listen_api_execution_pool = ThreadPoolExecutor(
```
self.listen_api_execution_pool is no more really used since self.execution_pool is no more used in ApiConsumer.
The global variable may add confusion, it would be clearer to remove self.list_api_execution_pool and provide self.execution_pool instead to ApiConsumer constructors

gileri replied:
Indeed, that was a pretty bad implementation here. Fixed, and we still use two separate execution queues to avoid head-of-line blocking if a callback URL is slow.

JustWats commented:
<img width="1489" height="972" alt="image" src="https://github.com/user-attachments/assets/3a55849b-64d0-4bb6-a6f0-6c16807eb6a9" />

Just an update, @gileri been running stable about a few days with the changes. Testing a bunch of connectors at the moment to test thoroughput.

<img width="1171" height="583" alt="image" src="https://github.com/user-attachments/assets/84599c66-8828-4358-9784-4c3a7e768130" />

Running on 30-30-3 settings at the moment, and it feels pretty good.

<img width="1751" height="845" alt="image" src="https://github.com/user-attachments/assets/598b8d00-29d5-4343-946f-1c06f6cdebe3" />


Running on 10 node cluster for OCTI. 2 frontend, 2 ingestion, 5 backend, 1 api. 3 Workers per backend.
Running on elasticsearch cluster 3-3 master-data.  13Gi JVM heaps, redis-cluster with 25Gi RAM.

gileri commented:
Thanks for the reviews and suggestions @xfournet!

I've cleaned up the thread pools handling, and keep the two separate. Also reverted the pool size reduction: Performance would increase up to 5 times on single-queue workloads, but on systems with 5 or more active queues it would be only about 5-10% (AMQP prefetch added, 25ms wait average wait between message per queue removed).

Amending the PR description to avoid confusion.

xfournet commented on opencti-worker/src/worker.py:
```diff
@@ -493,16 +490,13 @@ def run(self) -> None:
                         method.delivery_tag,
                         data,
                     )
-                    while task_future.running():  # Loop while the thread is processing
-                        self.pika_connection.sleep(0.05)
-                    self.worker_logger.info("Message processed, thread terminated")
```
I identified a potential concurrency issue that we had not initially considered. Indeed, this change allows multiple messages per queue to be processed concurrently. In the Consumer class, the OpenCTI API client (self.api) is modified at the beginning of data_handler (lines 320 to 329) and used later (lines 346, 372, 403, 424, and 447), which leaves room for a race condition between concurrent threads.
Note: In ApiConsumer, this appears to be safe.

gileri commented:
> bundle in split with an order that should not be changed or if each small bundle are independent. So the part to check for concurrency is in this part https://github.com/OpenCTI-Platform/client-python/blob/a784e89fb0aba21a6cbba009ff457b0af65c29c4/pycti/utils/opencti_stix2_splitter.py#L204

[RabbitMQ Queues are FIFO, and this bundle-split system requeues sub-bundles in independent AMQP messages](https://github.com/OpenCTI-Platform/opencti/blob/f2053d0b0e38ba79b31089e3656d52b2d6fad20b/opencti-worker/src/worker.py#L369-L390), so order of processing for those is already not guaranteed nor respected. And the split operation itself seem pretty safe, now that HTTP headers are correctly sent.

>Could you target release/current branch for performance PRs instead of master please ? This will let more time to test on internal validation platform before releasing.

Done, thanks.

gileri commented on opencti-worker/src/worker.py:
```diff
@@ -316,17 +316,28 @@ def data_handler(  # pylint: disable=too-many-statements, too-many-locals
+                json_logging=self.json_logging,
+                perform_health_check=False,
+            )
+            api.logger_class("worker")
```
I didn't check the implications of not calling that, but [it seems that it can be safely removed](https://github.com/OpenCTI-Platform/client-python/blob/master/pycti/utils/opencti_logger.py).

gileri commented on opencti-worker/src/worker.py:
```diff
@@ -316,17 +316,28 @@ def data_handler(  # pylint: disable=too-many-statements, too-many-locals
         imported_items = []
         start_processing = datetime.datetime.now()
         try:
+            api = OpenCTIApiClient(
+                url=self.opencti_url,
+                token=self.opencti_token,
+                log_level=self.log_level,
+                ssl_verify=self.ssl_verify,
+                json_logging=self.json_logging,
+                perform_health_check=False,
+            )
```
Yep, I started implementing a pool for that, but [I've benchmarked it, it takes about 30ms CPU time per 1 000 messages](https://github.com/OpenCTI-Platform/opencti/pull/11849/commits/a3970b0c06aac5f7ac26a821001e89158de5148a). The tradeoff seems acceptable for the simplicity.

We would have needed a pool if we couldn't disable `perform_health_check`, but we can. An healthcheck is already done at worker start up, like before. And if platform becomes unavailable later, nothing changes how it's handled.

gileri commented:
I've done 2 full import of the MITRE dataset with original worker code, and two with the code in the PR.
Both result in the same number of OCTI entities created. Running multiple instances of the worker did not negatively impact the test either.

gileri commented:
[This test](https://github.com/OpenCTI-Platform/opencti/blob/b8595a371aad1879ff0c2af6446fce97a0843725/opencti-platform/opencti-graphql/tests/03-streams/00-Raw/raw-test.js#L63-L66) fails with the changes, due to 4 label edits present in stream instead of the expect 1 defined [here](https://github.com/OpenCTI-Platform/opencti/blob/b8595a371aad1879ff0c2af6446fce97a0843725/opencti-platform/opencti-graphql/tests/utils/syncCountHelper.ts#L56-L88).

However I see 4 labelEdit mutations [here](https://github.com/OpenCTI-Platform/opencti/blob/b8595a371aad1879ff0c2af6446fce97a0843725/opencti-platform/opencti-graphql/tests/02-integration/02-resolvers/label-test.js). What is the expected number here ? Should the syncCountHelper.ts file be updated ?

gileri commented:
I've updated the value, tests pass. Ready for review.

gileri commented:
Thanks @xfournet for the review!

xfournet commented on opencti-worker/src/message_queue_consumer.py:
```diff
@@ -15,12 +15,13 @@ class MessageQueueConsumer:  # pylint: disable=too-many-instance-attributes
         self.pika_connection = pika.BlockingConnection(self.pika_parameters)
         self.channel = self.pika_connection.channel()
-        self.channel.basic_qos(prefetch_count=1)
+        self.channel.basic_qos(prefetch_count=(self.execution_pool._max_workers + 1) if self.queue_concurrency_enabled else 1)
```
the +1 is needed to be ensure to have always a message ready to be consumed by the next worker thread.
Currently there is a roundtrip with RabbitMQ for each consumed message.

xfournet commented on opencti-worker/src/message_queue_consumer.py:
```diff
@@ -83,9 +84,10 @@ def consume_queue(self) -> None:
+                if not self.queue_concurrency_enabled:
+                    while task_future.running():  # Loop while the thread is processing
+                        self.pika_connection.sleep(0.05)
+                    self.logger.info("Message processed, thread terminated")
```
you're probably right, but i wanted to minimize the impact of the changes when the settings is not enabled. Especially in regards of flaky test, i notice that every changes around that introduce a different behavior on sync test events count.
i propose to reassess the need of this active wait later
